src/mlpack/core/metrics/bleu_impl.hpp

```diff
@@ -0,0 +1,186 @@
+  const bool smooth)
+{
+  // WordVector is a string container type.
+  // Also, TranslationCorpusType is an array of such containers.
+  typedef typename TranslationCorpusType::value_type WordVector;
+
+  // matchesByOrder: It catches how many times sequence of a particular order
+  // is encountered in both reference corpus and translation corpus.
+  std::vector<size_t> matchesByOrder(maxOrder, 0);
+
+  // possibleMatchesByOrder: It tracks how many possible matches can be in the
+  // translation corpus.
+  std::vector<size_t> possibleMatchesByOrder(maxOrder, 0);
+
+  // referenceLength: It is the sum of minimum length of the paragraph from
+  // various documents.
+  // translationLength: It is the sum of length of each paragraphs.
+  referenceLength = 0, translationLength = 0;
+
+  auto refIt = referenceCorpus.cbegin();
+  auto trIt = translationCorpus.cbegin();
+  for (; refIt != referenceCorpus.cend(), trIt != translationCorpus.cend();
+      ++refIt, ++trIt)
+  {
+    size_t min = std::numeric_limits<size_t>::max();
+    for (auto t : *refIt)
+    {
+      if (min > t.size())
+      {
+        min = t.size();
+      }
+    }
+    referenceLength += min;
+    translationLength += trIt->size();
+
+    // mergedRefNGramCounts: It accumulates all the similar n-grams from
+    // various references or documents, so that there is no repetition of
+    // any key (sequence of order n).
+    std::map<WordVector, size_t> mergedRefNGramCounts;
+    for (auto t : *refIt)
+    {
+      // ngram: It holds the n-grams of each document/reference.
+      const std::map<WordVector, size_t> ngrams = GetNGrams(t);
+      for (auto it = ngrams.cbegin(); it != ngrams.cend(); ++it)
+      {
+        if (!mergedRefNGramCounts[it->first])
+          mergedRefNGramCounts[it->first] = it->second;
+        else
+        {
+          mergedRefNGramCounts[it->first]
+              = std::max(mergedRefNGramCounts[it->first], it->second);
+        }
+      }
+    }
+    // translationNGramCounts: It extracts the n-grams of the generated text
+    // sequence.
+    std::map<WordVector, size_t> translationNGramCounts = GetNGrams(*trIt);
+
+    // overlap: It holds those key (sequence of order n) which are common to
+    // reference corpus and translation corpus.
+    std::map<WordVector, size_t> overlap;
+    for (auto it = mergedRefNGramCounts.cbegin();
+         it != mergedRefNGramCounts.cend();
+         ++it)
+    {
+        if (translationNGramCounts[it->first])
+        {
+          // If the key (sequence of order n) is present in both translation
+          // corpus as well as reference corpus, then the minimum number of
+          // counts it has occurred in any is considered.
+          overlap[it->first] = std::min(translationNGramCounts[it->first],
+                                        it->second);
+        }
+    }
+
+    for (auto it = overlap.cbegin(); it != overlap.cend(); ++it)
+    {
+      matchesByOrder[it->first.size() - 1] += it->second;
```

Comment:
Can the size be equal to zero? Do we need to check it?

Reply:
The `overlap` contains either no elements or contains elements with size at least `1`, and upto `maxOrder`. Because this is the length of the sequence of order `1` or `2` ..., and till `maxOrder`. It simply won't scan sequences of size `0`.

And if overlap itself has size `0`, then it won't enter inside the for loop.
I think it's not needed. Let me know what do you think?

Reply:
Yes, in this case the check isn't needed.